File: 0097-interleaving-string/0097-interleaving-string.java
class Solution {
    private Boolean[][][] dp;
    public boolean check(String s1,String s2,String s3,int index,int p1,int p2){
        if( index==s3.length() ) return true;
        
        if( p1==s1.length() && p2==s2.length()) return false;
        //-----------------------------------
        if(dp[index][p1][p2]!=null){
            return dp[index][p1][p2];
        }
        //-----------------------------
        if(p1==s1.length()){
            return (s2.charAt(p2) == s3.charAt(index)) && check(s1,s2,s3,index+1,p1,p2+1);
        }
        if(p2==s2.length()){
            return (s1.charAt(p1) == s3.charAt(index)) && check(s1,s2,s3,index+1,p1+1,p2);
        }
        
        
        boolean case1 =  (s1.charAt(p1) == s3.charAt(index)) && check(s1,s2,s3,index+1,p1+1,p2);

        boolean case2 =  (s2.charAt(p2) == s3.charAt(index)) && check(s1,s2,s3,index+1,p1,p2+1);
        
        dp[index][p1][p2] = case1 || case2;
        return case1 || case2;
        
        
    }
    public boolean isInterleave(String s1, String s2, String s3) {
        if(s1.length()+s2.length() != s3.length()){
            return false;
        }
        dp = new Boolean[s3.length()+1][s1.length()+1][s2.length()+1];
        
        return check(s1,s2,s3,0,0,0);
    }
}

File: 0097-interleaving-string/0097-interleaving-string.py
class Solution:
    @lru_cache(None)
    def check(self,index,p1,p2):
        
        #------------------
        s1 = self.s1 
        s2 = self.s2 
        s3 = self.s3 
        #--------------------
        if index==len(s3):  
            return True
        
        if p1==len(s1) and p2==len(s2): 
            return False
        
        if p1==len(s1):
            return (s2[p2] == s3[index])  and self.check(index+1,p1,p2+1)
        
        if p2==len(s2):
            return (s1[p1] == s3[index])  and self.check(index+1,p1+1,p2)
        
        
        
        case1 =  (s1[p1] == s3[index]) and self.check(index+1,p1+1,p2)
        case2 =  (s2[p2] == s3[index]) and self.check(index+1,p1,p2+1)
        
        
        return case1 or case2;
        
        
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        self.s1 = s1
        self.s2 = s2
        self.s3 = s3
        return len(s1)+len(s2)==len(s3) and self.check(0,0,0);